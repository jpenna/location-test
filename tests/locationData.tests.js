/* eslint import/no-extraneous-dependencies: "off" */
/*  eslint import/no-unresolved: "off" */

import React from 'react';
import PropTypes from 'prop-types';
import { render, mount, shallow } from 'enzyme';
import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import LocationData from 'src/components/locationData';

const axiosMock = new MockAdapter(axios);

describe('<LocationData />', () => {
  let rendered;
  let props;
  let url;
  let axiosResponse;

  beforeAll(() => {
    props = {
      title: 'title',
      locationData: {},
      setLocationData: jest.fn(),
      type: 'user',
    };
    url = 'ok.com';
    axiosResponse = { data: 'data' };

    rendered = render(<LocationData {...props} />);
  });

  beforeEach(() => {
    axiosMock.onGet(`http://freegeoip.net/json/${url}`)
      .reply(200, axiosResponse);
  });

  afterEach(() => {
    jest.clearAllMocks();
    axiosMock.reset();
  });

  afterAll(() => {
    axiosMock.restore();
  });

  it('should have H2 title', () => {
    expect(rendered.find('h2').text()).toEqual('title');
  });

  it('should have a button to fetch user location', () => {
    expect(rendered.find('button.get-data').length).toEqual(1);
  });

  it('should have a button to reset user location', () => {
    expect(rendered.find('button.reset').length).toEqual(1);
  });

  it('should return formatted list of location data', () => {
    const data = { user: 'me', null: '', how: 'what' };
    const list = LocationData.renderLocationData(data);
    const listRendered = render(<span>{list}</span>);
    expect(listRendered.find('div')).toHaveLength(2);
    expect(listRendered.html().search('user') > 0).toBeTruthy();
    expect(listRendered.html().search('what') > 0).toBeTruthy();
  });

  it('should get location automatically if URL is defined', () => {
    const fetcher = jest.spyOn(LocationData.prototype, 'getLocation');
    const mounted = mount(<LocationData {...props} />);
    mounted.setProps({ url });
    expect(fetcher).toHaveBeenCalledTimes(1);
    fetcher.mockRestore();
  });

  it('getLocation() should call props.setLocation with type and data', () => {
    const setLocationData = jest.fn();
    const context = {
      props: { url, setLocationData, type: 'ops' },
      setState: jest.fn(),
    };
    return LocationData.prototype.getLocation.call(context)
      .then(() => {
        expect(setLocationData.mock.calls).toEqual([[context.props.type, axiosResponse]]);
        expect(context.setState.mock.calls).toEqual([[{ error: '' }]]);
      });
  });

  it('getLocation() should set error on exception', () => {
    const urlError = 'error';
    axiosMock.onGet(`http://freegeoip.net/json/${urlError}`)
      .networkError();
    const context = {
      setState: jest.fn(),
      props: { url: urlError, setLocationData: () => {} },
    };
    return LocationData.prototype.getLocation.call(context)
      .then(() => {
        expect(context.setState.mock.calls).toEqual([[{ error: 'Sorry, couldn\'t fecth data.' }]]);
      });
  });

  it('should reset location data when RESET button is clicked', () => {
    const shallowed = shallow(<LocationData {...props} />);
    shallowed.find('button.reset').simulate('click');
    expect(props.setLocationData).toHaveBeenCalledTimes(1);
  });

  it('should hide GET DATA button if props says so', () => {
    const renderedHidden = render(<LocationData {...props} hideButton />);
    expect(renderedHidden.find('button.get-data').parent().attr('hidden')).toEqual('hidden');
  });

  it('should hide RESET button if props says so', () => {
    const renderedHidden = render(<LocationData {...props} hideButton />);
    expect(renderedHidden.find('button.reset').attr('hidden')).toEqual('hidden');
  });

  it('should have the right TypeProps', () => {
    const propTypes = {
      url: PropTypes.string,
      hideButton: PropTypes.bool,
      title: PropTypes.string.isRequired,
      setLocationData: PropTypes.func.isRequired,
      locationData: PropTypes.object, // eslint-disable-line react/forbid-prop-types
      type: PropTypes.string.isRequired,
    };
    expect(LocationData.propTypes).toEqual(propTypes);
  });

  it('should have the right DefaultProps', () => {
    const defaultProps = {
      url: '',
      hideButton: false,
      locationData: {},
    };
    expect(LocationData.defaultProps).toEqual(defaultProps);
  });
});
