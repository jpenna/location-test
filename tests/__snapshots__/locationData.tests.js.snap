// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationData /> should render correctly 1`] = `
<div>
  <div>
    <div
      className="container"
    >
      <div
        className="notification is-primary info-balloon is-hidden"
      >
        <button
          className="delete"
          onClick={[Function]}
        />
        This is your data according to freejeoip.net at 
        
      </div>
      <button
        className="js-showInfo is-buttonless "
        disabled={true}
        onClick={[Function]}
      >
        <h3
          className="title is-5 is-inline-block"
        >
          title
        </h3>
        <span
          className="icon info-sign is-hidden"
        >
          <i
            className="fa fa-info-circle"
          />
        </span>
      </button>
      <span
        className="field"
        hidden={false}
      >
        <button
          className="reset-button button is-small is-borderless is-danger inverted text-bottom-align"
          onClick={[Function]}
          type="button"
        >
          <span
            className="icon"
          >
            <i
              className="fa fa-trash"
            />
          </span>
        </button>
      </span>
    </div>
  </div>
  <div
    className="section"
    hidden={1}
  >
    <button
      className="get-location-button is-buttonless pointer"
      onClick={[Function]}
    >
      <div
        className="icon button is-info is-rounded margin-bottom-half"
      >
        <i
          className="fa fa-map-marker"
        />
      </div>
      <div
        className="title is-6"
      >
        My Location
      </div>
    </button>
  </div>
  <div
    className="section"
    hidden={true}
  >
    <div
      className="content has-text-centered"
      hidden=""
    >
      Let's get some page info?
      <br />
      Type in a URL above
    </div>
  </div>
  <div
    className="is-danger has-text-centered help title is-6 has-new-line is-marginless"
  >
    
  </div>
  <div
    hidden=""
  />
</div>
`;
